javascript/ql/src/semmle/javascript/RangeAnalysis.qll

```diff
@@ -0,0 +1,366 @@
+    exists(Expr left, Expr right, int bias1, int bias2 | left = comparison.getLeftOperand() and right = comparison.getRightOperand() |
+      // A <= B + c
+      linearDefinition(left.flow(), a, asign, bias1) and
+      linearDefinition(right.flow(), b, bsign, bias2) and
+      bias = bias2 - bias1
+      or
+      // A - B + c1 <= c2  becomes  A <= B + (c2 - c1)
+      linearDefinitionSum(left.flow(), a, asign, b, -bsign, bias1) and
+      right.getIntValue() = bias2 and
+      bias = bias2 - bias1
+      or
+      // c1 <= -A + B + c2  becomes  A <= B + (c2 - c1)
+      left.getIntValue() = bias1 and
+      linearDefinitionSum(right.flow(), a, -asign, b, bsign, bias2) and
+      bias = bias2 - bias1
+    )
+  }
+
+  /**
+   * Holds if `guard` asserts that the outcome of `A <op> B + bias` is true, where `<op>` is a comparison operator.
+   */
+  predicate linearComparisonGuard(ConditionGuardNode guard, DataFlow::Node a, int asign, string operator, DataFlow::Node b, int bsign, int bias) {
+    exists (Comparison compare | compare = getDefinition(guard.getTest().flow()).asExpr() |
+      linearComparison(compare, a, asign, b, bsign, bias) and
+      (
+        guard.getOutcome() = true and operator = compare.getOperator()
+        or
+        guard.getOutcome() = false and operator = negateOperator(compare.getOperator())
+      )
+    )
+  }
+
+  /**
+   * Gets the binary operator whose return value is the opposite of `operator` (excluding NaN comparisons).
+   */
+  private string negateOperator(string operator) {
+    operator = "==" and result = "!=" or
+    operator = "!=" and result = "==" or
+    operator = "===" and result = "!==" or
+    operator = "!==" and result = "===" or
+    operator = "<" and result = ">=" or
+    operator = "<=" and result = ">" or
+    operator = ">" and result = "<=" or
+    operator = ">=" and result = "<"
+  }
+
+  /**
+   * Holds if immediately after `cfg` it becomes known that `A <= B + c`.
+   *
+   * These are the initial inputs to the difference bound constraint system.
+   *
+   * The dual constraint `-B <= -A + c` is not included in this predicate.
+   */
+  predicate comparisonEdge(ControlFlowNode cfg, DataFlow::Node a, int asign, DataFlow::Node b, int bsign, int bias) {
+    // A <= B + c
+    linearComparisonGuard(cfg, a, asign, "<=", b, bsign, bias)
+    or
+    // A <= B + c   iff   A < B + c + 1  (assuming A,B are integers)
+    linearComparisonGuard(cfg, a, asign, "<", b, bsign, bias + 1)
+    or
+    // A <= B + c   iff   B >= A - c
+    linearComparisonGuard(cfg, b, bsign, ">=", a, asign, -bias)
+    or
+    // A <= B + c   iff   B > A - c - 1  (assuming A,B are integers)
+    linearComparisonGuard(cfg, b, bsign, ">", a, asign, -bias - 1)
+    or
+    exists (string operator | operator = "==" or operator = "===" |
+      // A == B + c   iff   A <= B + c  and  B <= A - c
+      linearComparisonGuard(cfg, a, asign, operator, b, bsign, bias)
+      or
```

Comment:
The left-hand of the `iff` is the expression we are matching. The right-hand are the two constraints we are generating. Generating two constraints means using an `or` because the predicate has to hold for either of the two constraints.